### PR TITLE
Add precompile entries autogenerated by SnoopCompile

### PR DESCRIPTION
**Pull request by joehuchette**
cc @mlubin @IainNZ

### CODE REVIEW

**Comment by mlubin** [comment]
Any difference on the benchmarks?

**Comment by joehuchette** [comment]
I'm getting a recompilation error in ForwardDiff, so I'm still not sure if it works at all.

File: src/precompile.jl [hunk 1]
```diff
@@ -115,3 +115,255 @@ Base.precompile(parseExpr, (Float64,Symbol,Vector{Any},Vector{Any},Symbol))
...
+precompile(JuMP.addToExpression, (JuMP.GenericAffExpr{Float64, JuMP.Variable}, Float64, JuMP.Variable,))
+precompile(JuMP.setObjectiveSense, (JuMP.Model, Symbol,))
+precompile(JuMP._sizehint_expr!, (JuMP.GenericAffExpr{Float64, JuMP.Variable}, Int64,))
+precompile(JuMP.addToExpression, (JuMP.GenericAffExpr{Float64, JuMP.Variable}, Float64, JuMP.GenericQuadExpr{Float64, JuMP.Variable},))
+precompile(JuMP.call, (Type{JuMP.ProblemTraits}, Bool, Bool, Bool, Bool, Bool, Bool, Bool, Bool, Bool,))
+precompile(JuMP.setObjective, (JuMP.Model, Symbol, JuMP.GenericAffExpr{Float64, JuMP.Variable},))
+precompile(JuMP.resize!, (JuMP.IndexedVector{Float64}, Int64,))
+precompile(JuMP.assert_validmodel, (Expr, Expr,))
+precompile(JuMP.parseExprToplevel, (Symbol, Symbol,))
+precompile(JuMP.__init__, ())
```
Comment (IainNZ) [inline src/precompile.jl]: This seems sketchy

File: src/precompile.jl [hunk 2]
```diff
@@ -115,3 +115,255 @@ Base.precompile(parseExpr, (Float64,Symbol,Vector{Any},Vector{Any},Symbol))
...
+precompile(JuMP.parseNLExpr, (Expr, Symbol, Symbol, Int64, Symbol,))
+precompile(JuMP.parseNLExpr, (Expr, Symbol, Symbol, Symbol, Symbol,))
+precompile(JuMP.parseNLExpr, (Expr, Symbol, Int64, Int64, Symbol,))
+precompile(JuMP.parseNLExpr, (Int64, Symbol, Symbol, Int64, Symbol,))
+precompile(JuMP.isdependent, (Array{Any, 1}, Expr, Int64,))
+precompile(JuMP.buildrefsets, (Void,))
+precompile(JuMP.tryParseIdxSet, (Expr,))
+precompile(JuMP.parseIdxSet, (Expr,))
+precompile(JuMP.assert_validmodel, (Expr, Expr,))
+precompile(JuMP.__init__, ())
```
Comment (IainNZ) [inline src/precompile.jl]: I guess this from multiple runs?

Comment (joehuchette) [inline src/precompile.jl]: Yeah, I ran it for each of the benchmarks and concatenated them.

**Comment by joehuchette** [comment]
Output for ForwardDiff and ReverseDiffSparse, if you're interested:
```
huchette@this-is-not-a-computer:/tmp/precompile$ cat precompile_ForwardDiff.jl
function _precompile_()
    ccall(:jl_generating_output, Cint, ()) == 1 || return nothing
    precompile(ForwardDiff.tupexpr, (Function, Int64,))
    precompile(ForwardDiff.zero_tuple, (Vararg{Any},))
    precompile(ForwardDiff.mul_tuples, (Vararg{Any},))
    precompile(ForwardDiff.minus_tuple, (Vararg{Any},))
    precompile(ForwardDiff.scale_tuple, (Vararg{Any},))
    precompile(ForwardDiff.add_tuples, (Vararg{Any},))
end

huchette@this-is-not-a-computer:/tmp/precompile$ cat precompile_ReverseDiffSparse.jl
function _precompile_()
    ccall(:jl_generating_output, Cint, ()) == 1 || return nothing
    precompile(ReverseDiffSparse.forward_eval, (Array{Float64, 1}, Array{Float64, 1}, Array{ReverseDiffSparse.NodeData, 1}, Base.SparseMatrix.SparseMatrixCSC{Bool, Int64}, Array{Float64, 1}, Array{Float64, 1}, Array{Float64, 1}, Array{Float64, 1}, Array{Any, 1}, Array{Any, 1},))
    precompile(ReverseDiffSparse.forward_eval, (Array{Float64, 1}, Array{Float64, 1}, Array{ReverseDiffSparse.NodeData, 1}, Base.SparseMatrix.SparseMatrixCSC{Bool, Int64}, Array{Float64, 1}, Array{Float64, 1}, Array{Float64, 1}, Array{Float64, 1}, Array{Float64, 1}, Array{Float64, 1},))
    precompile(ReverseDiffSparse.Coloring.recovery_preprocess, (ReverseDiffSparse.Coloring.UndirectedGraph, Array{Int64, 1}, Int64, Array{Int64, 1},))
    precompile(ReverseDiffSparse.compute_hessian_sparsity, (Array{ReverseDiffSparse.NodeData, 1}, Base.SparseMatrix.SparseMatrixCSC{Bool, Int64}, Array{ReverseDiffSparse.Linearity, 1}, Array{Base.Set{Tuple{Int64, Int64}}, 1}, Array{Base.Set{Int64}, 1},))
    precompile(ReverseDiffSparse.Coloring.gen_adjlist, (Array{Int64, 1}, Array{Int64, 1}, Int64,))
    precompile(ReverseDiffSparse.order_subexpressions, (Array{Array{ReverseDiffSparse.NodeData, 1}, 1}, Array{Array{ReverseDiffSparse.NodeData, 1}, 1},))
    precompile(ReverseDiffSparse.classify_linearity, (Array{ReverseDiffSparse.NodeData, 1}, Base.SparseMatrix.SparseMatrixCSC{Bool, Int64}, Array{ReverseDiffSparse.Linearity, 1}, Base.BitArray{1},))
    precompile(ReverseDiffSparse.Coloring.acyclic_coloring, (ReverseDiffSparse.Coloring.UndirectedGraph,))
    precompile(ReverseDiffSparse.Coloring.hessian_color_preprocess, (Base.Set{Tuple{Int64, Int64}}, Int64,))
    precompile(ReverseDiffSparse.Coloring.indirect_recover_structure, (ReverseDiffSparse.Coloring.RecoveryInfo,))
    precompile(ReverseDiffSparse.simplify_constants, (Array{Float64, 1}, Array{ReverseDiffSparse.NodeData, 1}, Base.SparseMatrix.SparseMatrixCSC{Bool, Int64}, Array{Float64, 1}, Array{ReverseDiffSparse.Linearity, 1},))
    precompile(ReverseDiffSparse.Coloring.depth_first_visit_impl!, (Array{Int64, 1}, Array{Int64, 1}, Array{Int64, 1}, Array{Int64, 1}, Array{Int64, 1}, ReverseDiffSparse.Coloring.TopologicalSortVisitor,))
    precompile(ReverseDiffSparse.has_user_multivariate_operators, (Array{ReverseDiffSparse.NodeData, 1},))
    precompile(ReverseDiffSparse.Coloring.prepare_seed_matrix!, (Array{Float64, 2}, ReverseDiffSparse.Coloring.RecoveryInfo,))
    precompile(ReverseDiffSparse.reverse_eval, (Array{Float64, 1}, Array{Float64, 1}, Array{ReverseDiffSparse.NodeData, 1}, Base.SparseMatrix.SparseMatrixCSC{Bool, Int64},))
    precompile(ReverseDiffSparse.compute_gradient_sparsity, (Array{ReverseDiffSparse.NodeData, 1},))
    precompile(ReverseDiffSparse.reverse_extract, (Array{Float64, 1}, Array{Float64, 1}, Array{ReverseDiffSparse.NodeData, 1}, Base.SparseMatrix.SparseMatrixCSC{Bool, Int64}, Array{Float64, 1}, Float64,))
    precompile(ReverseDiffSparse.list_subexpressions, (Array{ReverseDiffSparse.NodeData, 1},))
    precompile(ReverseDiffSparse.Coloring.reverse_topological_sort_by_dfs, (Array{Int64, 1}, Array{Int64, 1}, Int64, Array{Int64, 1}, Array{Int64, 1}, Array{Int64, 1},))
    precompile(ReverseDiffSparse.Coloring.call, (Type{ReverseDiffSparse.Coloring.TopologicalSortVisitor}, Int64,))
    precompile(ReverseDiffSparse.adjmat, (Array{ReverseDiffSparse.NodeData, 1},))
    precompile(ReverseDiffSparse.Coloring.call, (Type{ReverseDiffSparse.Coloring.RecoveryInfo},))
    precompile(ReverseDiffSparse.Coloring.close_vertex!, (ReverseDiffSparse.Coloring.TopologicalSortVisitor, Int64,))
    precompile(ReverseDiffSparse.Coloring.traverse_graph, (Array{Int64, 1}, Array{Int64, 1}, Int64, ReverseDiffSparse.Coloring.TopologicalSortVisitor, Array{Int64, 1}, Array{Int64, 1}, Array{Int64, 1},))
end
```

**Comment by mlubin** [comment]
Still getting errors in ForwardDiff? What are they?

**Comment by mlubin** [comment]
A few of the signatures in ReverseDiffSparse are about to change from my speed tweaks: https://github.com/mlubin/ReverseDiffSparse.jl/tree/speedup

**Comment by joehuchette** [comment]
I'm seeing a speedup of ~4.5 seconds on fac2000 on my machine

**Comment by joehuchette** [comment]
(that's with jh/precompile on Gurobi as well)

**Comment by mlubin** [comment]
Should we also be including ``ccall(:jl_generating_output, Cint, ()) == 1 || return nothing`` in the precompile files? Might be there for a reason.

**Comment by mlubin** [comment]
4.5 seconds is great

**Comment by joehuchette** [comment]
Seems like ~3.5 seconds on opf662, but I'm hitting 
```
ERROR: LoadError: UndefVarError: has_user_multivariate_operators not defined
 in call at /Users/huchette/.julia/v0.4/JuMP/src/nlp.jl:143
 in _buildInternalModel_nlp at /Users/huchette/.julia/v0.4/JuMP/src/nlp.jl:1133
 in buildInternalModel at /Users/huchette/.julia/v0.4/JuMP/src/solvers.jl:307
 in solve at /Users/huchette/.julia/v0.4/JuMP/src/solvers.jl:131
 in opf at /Users/huchette/Downloads/JuMPSupplement/acpower/opf.jl:170
 in include at /Applications/Julia-0.4.3.app/Contents/Resources/julia/lib/julia/sys.dylib
 in include_from_node1 at /Applications/Julia-0.4.3.app/Contents/Resources/julia/lib/julia/sys.dylib
 in process_options at /Applications/Julia-0.4.3.app/Contents/Resources/julia/lib/julia/sys.dylib
 in _start at /Applications/Julia-0.4.3.app/Contents/Resources/julia/lib/julia/sys.dylib
while loading /Users/huchette/Downloads/JuMPSupplement/acpower/opf.jl, in expression starting on line 239
```
pretty early on ¯\_(ツ)_/¯ 

The ccall couldn't hurt, I can add it back in.

**Comment by mlubin** [comment]
``has_user_multivariate_operators`` is defined in the latest release of ReverseDiffSparse

**Comment by joehuchette** [comment]
It strangely seems to speed up acpower ~3 to get to the ``solving`` line, but slow down the entire script considerably (44 seconds on acpower66200).

**Comment by mlubin** [comment]
44 seconds on 66200 is good, what was it before?

**Comment by joehuchette** [comment]
No, I'm saying precompilation _slows_ it down considerably. Are you benchmarking on che, or can I go and use that?

**Comment by mlubin** [comment]
Go ahead

**Comment by joehuchette** [comment]
I'm seeing maybe a second of speedup on lqcp and fac on che

**Comment by joehuchette** [comment]
Again, I'm seeing a substantial (~10 second) _slowdown_ using precompilation on the nonlinear benchmarks.

**Pull request closed without merging**